### PR TITLE
Add Erlang/OTP 24 to CI (+ bump ranch to 2.0.0 and adapt to it)

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
Opening as draft as per https://github.com/joedevivo/chatterbox/pull/145#issuecomment-878288845.

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter, I was able to debug the issue better, in your fork, and found the root cause (`ranch` 2.0.0 implements a different protocol callback, now solved). I had to `_checkouts` the thing, still, though, since it was eating the error up as `{error, undef}` and I couldn't understand what is was about.

If you could just unblock the CI I'd appreciate it much.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -15,15 +15,16 @@ jobs:
 
     strategy:
       matrix:
-        otp_version: ['23.2.1', '22.3.4.2']
+        otp_version: ['24', '23', '22']
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/main.yml:18]: I think this achieves the same effect as you wanted before, but we don't need to be so specific with `setup-beam`.

File: rebar.config [hunk 2]
```diff
@@ -13,8 +13,6 @@
 
 {ct_opts, [{verbose,true}]}.
 
-{dialyzer, [unknown]}.
```
Comment (paulo-ferraz-oliveira) [inline rebar.config:16]: Move to profile `test` as I wanted to benefit from the `deps` already there.

File: src/chatterbox_ranch_protocol.erl [hunk 3]
```diff
@@ -7,18 +7,18 @@
 %%-behaviour(ranch_protocol).
 
 -export([
-         start_link/4,
```
Comment (paulo-ferraz-oliveira) [inline src/chatterbox_ranch_protocol.erl:10]: All of this is `ranch` 2.0.0, except renaming `T` as `Transport` 😉

File: test/chatterbox_test_buddy.erl [hunk 4]
```diff
@@ -75,9 +75,8 @@ start(Config) ->
     {ok, _RanchPid} =
         ranch:start_listener(
           chatterbox_ranch_protocol,
-          10,
           ranch_ssl,
-          [{port, 8081}|proplists:get_value(ssl_options, Settings)],
+          #{ socket_opts => [{port, 8081}|proplists:get_value(ssl_options, Settings)] },
```
Comment (paulo-ferraz-oliveira) [inline test/chatterbox_test_buddy.erl:79]: `socket_opts` isn't actually required here, but this way it's more future-proof.

**Comment by paulo-ferraz-oliveira** [comment]
Tests pass on OTP 24, but fail on 23 (don't know about 22 yet). That's 😢.

**Comment by paulo-ferraz-oliveira** [comment]
As per https://github.com/paulo-ferraz-oliveira/chatterbox/runs/3047883585, stuff seems Ok.

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter, any way this gets merged soon? I don't mind porting it over to upstream, in any case.

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter, will this fork eventually get merged to upstream? Or do you intend this fork to be the "de facto" from now on? (asking so I can choose which one to better pull request to)

**Comment by tsloughter** [comment]
I plan to merge this upstream.

**Comment by paulo-ferraz-oliveira** [comment]
`rebar3_hank` also detects

```
test/http2c.erl:46: Field working_length in record http2c_state is unused
src/h2_stream.erl:64: Field promised_stream in record stream_state is unused
src/h2_stream.erl:63: Field next_state in record stream_state is unused
src/h2_stream.erl:62: Field response_end_stream in record stream_state is unused
src/h2_stream.erl:61: Field response_end_headers in record stream_state is unused
src/h2_stream.erl:57: Field request_end_headers in record stream_state is unused
src/h2_stream.erl:54: Field request_body in record stream_state is unused
src/h2_stream.erl:51: Field state in record stream_state is unused
src/h2_connection.erl:88: Field listen_ref in record connection is unused
src/h2_connection.erl:87: Field ssl_options in record connection is unused
config/sys.config:0: chatterbox_static_content_handler is not used anywhere in the code
```

Would trimming this be acceptable?

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter, any chance this gets merged soon?

**Comment by tsloughter** [comment]
I can merge today but suspect `cth_readable` shouldn't be a dep?

I take it you are using chatterbox with ranch 2.0? I've wanted to just remove the ranch module in order to not have it part of maintenance, esp since I use `acceptor_pool` instead, like in `grpcbox` it uses `acceptor_pool`. A reason I haven't done that is the ct suites currently use ranch to setup the server and it was a pain when I tried to replace that.

File: rebar.config [hunk 5]
```diff
@@ -24,7 +22,20 @@
             nowarn_export_all
            ]},
           {deps,
-           [{ranch, "1.2.1"}]}
+           [{ranch, "2.0.0"}, {cth_readable, "v1.5.1"}]}
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply rebar.config:25]: I might've been doing `rebar3 as test` at one point and it was complaining. I can't actually remember. Will remove.

[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply rebar.config]: Done.

**Comment by paulo-ferraz-oliveira** [comment]
I am using it with `ranch` 2.0, yes. And I get what you're saying, it was (albeit not a pain) the "hardest" part of this change.

**Comment by paulo-ferraz-oliveira** [comment]
What about what `hank` detects? Do you wanna trim any of the unused stuff?

**Comment by tsloughter** [comment]
Oh, I missed your hank comment. Yes, that would be good most likely. A separate PR I assume? I'll merge this one after CI finishes.

**Review by tsloughter — APPROVED** [review]

**Comment by tsloughter** [comment]
Argh, CI failed, I reran it in case it was just a blip.

**Comment by paulo-ferraz-oliveira** [comment]
> Argh, CI failed, I reran it in case it was just a blip.

It seems it's transient. I'll take a look at it once I find time.

**Comment by tsloughter** [comment]
Any updates on the tests? 

This will get more of my attention soon, I start working again on the 13th and part of that job will involve grpcbox and thus chatterbox.

**Comment by paulo-ferraz-oliveira** [comment]
Yeah, sorry, no updates on the tests. Was already happening to me using the previous version (prior to my pull request): sometimes they passed, sometimes they didn't. In any case, I don't think I'll have much time to work on this in the coming weeks, so feel free to proceed as you see fit.

**Comment by paulo-ferraz-oliveira** [comment]
Rebased on top of `master`. For reference:
* `rebar3 eunit` consistently passes all tests,
* `rebar3 ct --cover` (I assume this was the one failing before) passes locally,
so I know not how to proceed.

**Comment by paulo-ferraz-oliveira** [comment]
It passed for:
* https://github.com/paulo-ferraz-oliveira/chatterbox/runs/4809827060?check_suite_focus=true
* https://github.com/paulo-ferraz-oliveira/chatterbox/runs/4809828325?check_suite_focus=true

It failed for:
* https://github.com/paulo-ferraz-oliveira/chatterbox/runs/4809829407?check_suite_focus=true

❤️

**Comment by paulo-ferraz-oliveira** [comment]
These tests seem to consistently pass over OTP 24. I don't mind waiting for further changes to `master`, though, before I rebase this PR on top of it though. What say you?

Stuff:
* https://github.com/paulo-ferraz-oliveira/chatterbox/actions/runs/1695084076
* https://github.com/paulo-ferraz-oliveira/chatterbox/actions/runs/1695084742
* https://github.com/paulo-ferraz-oliveira/chatterbox/actions/runs/1695085104
* https://github.com/paulo-ferraz-oliveira/chatterbox/actions/runs/1695085301
* https://github.com/paulo-ferraz-oliveira/chatterbox/actions/runs/1695085568
(all these runs differ only on a white line - to be repeated)

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter, is there a way to move this forward? I don't mind closing the PR if you think it's not ready or brings no added value. Thanks.

**Comment by tsloughter** [comment]
Oh, sorry about that, merging now!

**Comment by paulo-ferraz-oliveira** [comment]
Cool, thanks.

**Pull request merged**